sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/SchedulerBuilder.java

```diff
@@ -0,0 +1,503 @@
+                        new PersisterCache(CuratorPersister.newBuilder(serviceSpec).build()) :
+                        CuratorPersister.newBuilder(serviceSpec).build());
+    }
+
+    SchedulerBuilder(ServiceSpec serviceSpec, SchedulerConfig schedulerConfig, Persister persister) {
+        this.serviceSpec = serviceSpec;
+        this.schedulerConfig = schedulerConfig;
+        this.persister = persister;
+    }
+
+    /**
+     * Returns the {@link ServiceSpec} which was provided via the constructor.
+     */
+    public ServiceSpec getServiceSpec() {
+        return serviceSpec;
+    }
+
+    /**
+     * Returns the {@link SchedulerFlags} object which was provided via the constructor.
+     */
+    public SchedulerConfig getSchedulerConfig() {
+        return schedulerConfig;
+    }
+
+    /**
+     * Specifies a custom {@link StateStore}, otherwise the return value of
+     * {@link DefaultScheduler#createStateStore(ServiceSpec, SchedulerFlags)} will be used.
+     *
+     * The state store persists copies of task information and task status for all tasks running in the service.
+     *
+     * @throws IllegalStateException if the state store is already set, via a previous call to either
+     * {@link #setStateStore(StateStore)} or to {@link #getStateStore()}
+     */
+    public SchedulerBuilder setStateStore(StateStore stateStore) {
+        if (stateStoreOptional.isPresent()) {
+            // Any customization of the state store must be applied BEFORE getStateStore() is ever called.
+            throw new IllegalStateException("State store is already set. Was getStateStore() invoked before this?");
+        }
+        this.stateStoreOptional = Optional.ofNullable(stateStore);
+        return this;
+    }
+
+    /**
+     * Returns the {@link StateStore} provided via {@link #setStateStore(StateStore)}, or a reasonable default
+     * created via {@link DefaultScheduler#createStateStore(ServiceSpec, SchedulerFlags)}.
+     *
+     * In order to avoid cohesiveness issues between this setting and the {@link #build()} step,
+     * {@link #setStateStore(StateStore)} may not be invoked after this has been called.
+     */
+    public StateStore getStateStore() {
+        if (!stateStoreOptional.isPresent()) {
+            setStateStore(new StateStore(persister));
+        }
+        return stateStoreOptional.get();
+    }
+
+    /**
+     * Specifies a custom {@link ConfigStore}.
+     *
+     * The config store persists a copy of the current configuration ('target' configuration),
+     * while also storing historical configurations.
+     */
+    public SchedulerBuilder setConfigStore(ConfigStore<ServiceSpec> configStore) {
+        if (configStoreOptional.isPresent()) {
+            // Any customization of the state store must be applied BEFORE getConfigStore() is ever called.
+            throw new IllegalStateException(
+                    "Config store is already set. Was getConfigStore() invoked before this?");
```

Comment:
This is meant to cover the case where someone does the following:

```
schedulerBuilder.getConfigStore(); // constructs and returns default ConfigStore
schedulerBuilder.setConfigStore(differentConfigStore); // throws because a default ConfigStore was already created
```

the intent is to throw if there's a potential for two different config stores, as in above scenario